### DISCUSSION TITLE
Livewire Loading property

### DISCUSSION THREAD

**Discussion by duncanmcclean** [livewire/livewire#1047] [4. General]
I'd like to display a loading screen to the user of my application when they are in between steps (I'm building a wizard-type app) because some of the steps have processing that needs done before moving onto the next step - eg. creating a 'checkout' in a payment gateway.

## What I've got so far
So I've created a public property called `loading` and I've set it to true right off the bat.

```php
public bool $loading = true;
```

Then in my component's `render` method, I set `loading` to false. To be honest, I could probably just set it to `false` to start with but that's away from the point.

When the user moves onto the next step, they click a button which has `wire:click="nextStep"` on it which triggers the `nextStep` method. That method is the one that calls the processing, but it's also the one where I set the loading properties.

```php
public function nextStep()
    {
        $this->clearValidation();
        
        $this->loading = true;

        if ($this->step === 2) {
            $this->createCheckout();
        }

        $this->step++;
        $this->loading = false;
}
```

Also, here's the bit in my Livewire component where I actually display the loading modal thing.

```
<div>
    @if($loading)
        <div class="fixed top-0 left-0 right-0 bottom-0 w-full h-screen z-50 overflow-hidden bg-gray-700 opacity-75 flex flex-col items-center justify-center">
            <img src="/static/img/loading.gif" alt="Page loading... please wait" width="40px">
        </div>
    @endif

    … the rest of my component
</div>
```

I've commented out the if statement and it shows so not an issue with the modal itself.

## Where I'm stuck
Basically, I'm not getting the loading modal to actually display. I later realised that I've been setting the `loading` property inside the same method, which obviously won't work (duh). So I commented out the `$loading = false` but it wouldn't show then.

So even when I set `loading` to `true` it stays `false`. I've tried setting it two different ways:
```php
$this->loading = true;
$this->fill(['loading' => false]);
```

Hopefully it's a simple solution, I'd love to know what I'm doing wrong.

**Comment by danmatthews** [comment]
I think it's because livewire runs the whole function in one round trip AJAX request.

So essentially loading = true will never be the case in your template.

**Reply by ryangjchandler** [reply]
Yeah, that's it. Livewire will run all of your actions and _then_ run the `render` method, so your template will receive `$loading` as `false`.

This is probably something you want to do either with JavaScript or see if you can work around it using `wire:loading`.

**Reply by danmatthews** [reply]
>  I later realised that I've been setting the loading property inside the same method, 

Looks like you caught this, and it's still not showing if you remove `$this->loading = false` ?

I would check that it's not clashing with the global `wire:loading` attribute by changing the name of the variable to something like `myLoading`.

**Reply by ryangjchandler** [reply]
I don't think Livewire actually sets any properties on the class when loading, since that stuff happens on the client-side when the AJAX request gets sent.

**Comment by duncanmcclean** [comment]
I've managed to fix it! 🎉 

I was over engineering my own solution as Livewire already provides a `wire:loading` thing which can be used for my exact purpose. 

Thanks for your help @danmatthews & @ryangjchandler!
